docs/provisioning.rst

```diff
@@ -0,0 +1,355 @@
+
+The kernel manager and kernel provisioner relationship
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Prior to this enhancement, the only extension point for customizing a
+kernel's behavior could occur by subclassing ``KernelManager``. This
+proved to be a limitation because the Jupyter framework allows for a
+single ``KernelManager`` class at any time. While applications could
+introduce a ``KernelManager`` subclass of their own, that
+``KernelManager`` was then tied directly to *that* application and
+thereby not usable as a ``KernelManager`` in another application. As a
+result, we consider the ``KernelManager`` class to be an
+*application-owned entity* upon which application-specific behaviors can
+be implemented.
+
+Kernel provisioners, on the other hand, are contained within the
+``KernelManager`` (i.e., a *has-a* relationship) and applications are
+agnostic as to what *kind* of provisioner is in use other than what is
+conveyed via the kernel's specification (kernelspec). All kernel
+interactions still occur via the ``KernelManager`` and ``KernelClient``
+classes within ``jupyter_client`` and potentially subclassed by the
+application.
+
+Kernel provisioners are not related in any way to the ``KernelManager``
+instance that controls their lifecycle, nor do they have any affinity to
+the application within which they are used. They merely provide a
+vehicle by which authors can extend the landscape in which a kernel can
+reside, while not side-effecting the application. That said, some kernel
+provisioners may introduce requirements on the application. For example
+(and completely hypothetically speaking), a ``SlurmProvisioner`` may
+impose the constraint that the server (``jupyter_client``) resides on an
+edge node of the Slurm cluster. These kinds of requirements can be
+mitigated by leveraging applications like `Jupyter Kernel Gateway <https://github.com/jupyter/kernel_gateway>`_ or
+`Jupyter Enterprise Gateway <https://github.com/jupyter/enterprise_gateway>`_
+where the gateway server resides on the edge
+node of (or within) the cluster, etc.
+
+Discovery
+~~~~~~~~~
+
+Kernel provisioning does not alter today's kernel discovery mechanism
+that utilizes well-known directories of ``kernel.json`` files. Instead,
+it optionally extends the current ``metadata`` stanza within the
+``kernel.json`` to include the specification of the kernel provisioner
+name, along with an optional ``config`` stanza, consisting of
+provisioner-specific configuration items. For example, a container-based
+provisioner will likely need to specify the image name in this section.
+The important point is that the content of this section is
+provisioner-specific.
+
+.. code:: JSON
+
+      "metadata": {
+        "kernel_provisioner": {
+          "provisioner_name": "k8s-provisioner",
+          "config": {
+              "image_name": "my_docker_org/kernel:2.1.5",
+              "max_cpus": 4
+          }
+        }
+      },
+
```

Comment:
Would it be possible to have an additional config_schemastanza that would contain a json schema (https://json-schema.org/) of the config?

That schema could be used by consumer application (eg jupyterlab to show a form in a UI).

Reply:
Yes, I agree, but feel this falls into the _kernel paramertization_ discussion, which should occur once this is merged.  I believe there will need to be two forms of schemas - one schema produced by the kernel itself describing kernel-specific parameters, the other schema produced by the provisioner describing provisioner-specific parameters.